pkg/validation/usage_groups.go

```diff
@@ -0,0 +1,187 @@
+	match := UsageGroupMatch{
+		tenantID: c.TenantID,
+	}
+
+	for name, matchers := range c.config {
+		if matchesAll(matchers, lbls) {
+			match.names = append(match.names, name)
+		}
+	}
+
+	return match
+}
+
+func (c *UsageGroupConfig) UnmarshalYAML(value *yaml.Node) error {
+	m := make(map[string]string)
+	err := value.DecodeWithOptions(&m, yaml.DecodeOptions{
+		KnownFields: true,
+	})
+	if err != nil {
+		return fmt.Errorf("malformed usage group config: %w", err)
+	}
+
+	*c, err = NewUsageGroupConfig(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *UsageGroupConfig) UnmarshalJSON(bytes []byte) error {
+	m := make(map[string]string)
+	err := json.Unmarshal(bytes, &m)
+	if err != nil {
+		return fmt.Errorf("malformed usage group config: %w", err)
+	}
+
+	*c, err = NewUsageGroupConfig(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type UsageGroupMatch struct {
+	tenantID string
+	names    []string
+}
+
+func (m UsageGroupMatch) CountReceivedBytes(profileType string, n int64) {
+	if len(m.names) == 0 {
+		usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, m.tenantID, noMatchName).Add(float64(n))
+		return
+	}
+
+	for _, name := range m.names {
+		usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, m.tenantID, name).Add(float64(n))
+	}
+}
+
+func (m UsageGroupMatch) CountDiscardedBytes(reason string, n int64) {
+	if len(m.names) == 0 {
+		usageGroupDiscardedBytes.WithLabelValues(reason, m.tenantID, noMatchName).Add(float64(n))
+		return
+	}
+
+	for _, name := range m.names {
+		usageGroupDiscardedBytes.WithLabelValues(reason, m.tenantID, name).Add(float64(n))
+	}
+}
+
+func NewUsageGroupConfig(m map[string]string) (UsageGroupConfig, error) {
+	if len(m) > maxUsageGroups {
+		return UsageGroupConfig{}, fmt.Errorf("maximum number of usage groups is %d, got %d", maxUsageGroups, len(m))
+	}
+
+	config := UsageGroupConfig{
+		config: make(map[string][]*labels.Matcher),
+	}
+
+	for name, matchersText := range m {
+		name = strings.TrimSpace(name)
+		if name == "" {
+			return UsageGroupConfig{}, fmt.Errorf("usage group name cannot be empty")
+		}
+
+		// TODO(bryanhuhta): We should probably validate the usage group name
+		// is a valid label value.
+
+		matchers, err := parser.ParseMetricSelector(matchersText)
+		if err != nil {
+			return UsageGroupConfig{}, fmt.Errorf("failed to parse matchers for usage group %q: %w", name, err)
+		}
+
+		config.config[name] = matchers
+	}
+
+	return config, nil
+}
+
+func (o *Overrides) DistributorUsageGroups(tenantID string) *UsageGroupConfig {
+	config := o.getOverridesForTenant(tenantID).DistributorUsageGroups
+
+	// It should never be nil, but check just in case!
+	if config == nil {
+		config = &UsageGroupConfig{}
+	}
+
+	// Make sure the usage group config has the tenant ID set. It is not known
+	// when it is unmarshaled from the configuration.
+	if config.TenantID == "" {
+		config.TenantID = tenantID
+	}
```

Comment:
This feels a bit sketchy to me, we despite not registering it as through "RegisterFlag" in theory you could still set in the pyroscope.yaml config a default usage groups for all tenants.

This bit would then expose the same *UsageGroupConfig to all tenant that don't override it, with a race to that bit which sets the tenant.

I think it would be better to do the tenant field in a wrapped type or even simpler put it to a parameter to the `Count*()` methods

```
type UsageGroupConfig // without TenantID

type UsageGroupInstance{
*UsageGroupConfig
tenantID string
}
```